hack/jenkins/test-history/gen_json.py

```diff
@@ -0,0 +1,191 @@
+    """Generates test data out of the provided JUnit path.
+
+    Returns None if there's an issue parsing the XML.
+    Yields name, time, failed, skipped for each test.
+    """
+    try:
+        data = subprocess.check_output(
+            ['gsutil', 'cat', path],
+            stderr=open(os.devnull, 'w'))
+    except subprocess.CalledProcessError:
+        return
+
+    try:
+        data = zlib.decompress(data, zlib.MAX_WBITS | 16)
+    except zlib.error:
+        # Don't fail if it's not gzipped.
+        pass
+    data = data.decode('utf-8')
+
+    try:
+        root = ET.fromstring(data)
+    except ET.ParseError:
+        return
+
+    for child in root:
+        name = child.attrib['name']
+        time = float(child.attrib['time'])
+        failed = False
+        skipped = False
+        for param in child:
+            if param.tag == 'skipped':
+                skipped = True
+            elif param.tag == 'failure':
+                failed = True
+        yield name, time, failed, skipped
+
+def get_tests_from_junit_path(path):
+    """Generates all tests in a JUnit GCS path."""
+    for test in gcs_get_tests(path):
+        if test is None:
+            continue
+        yield test
+
+def get_tests_from_build(job, build):
+    """Generates all tests for a build."""
+    for junit_path in gcs_ls_junit_paths(job, build):
+        for test in get_tests_from_junit_path(junit_path):
+            yield test
+
+def get_daily_builds(server, prefix):
+    """Generates all (job, build) pairs for the last day."""
+    now = time.time()
+    for job in get_jobs(server):
+        if not job.startswith(prefix):
+            continue
+        for build in reversed(sorted(get_builds(server, job))):
+            building, timestamp = get_build_info(server, job, build)
+            # Skip if it's still building.
+            if building:
+                continue
+            # Quit once we've walked back over a day.
+            if now - timestamp / 1000 > 60*60*24:
+                break
+            yield job, build
+
+def get_tests(server, prefix):
+    """Returns a dictionary of tests to be JSON encoded."""
+    tests = {}
+    for job, build in get_daily_builds(server, prefix):
+        print('{}/{}'.format(job, str(build)))
+        for name, duration, failed, skipped in get_tests_from_build(job, build):
+            if name not in tests:
+                tests[name] = {}
+            if skipped:
+                continue
+            if job not in tests[name]:
+                tests[name][job] = []
+            tests[name][job].append({
+                'build': build,
+                'failed': failed,
+                'time': duration
+            })
+    return tests
+
+if __name__ == '__main__':
+    if len(sys.argv) != 3:
+        print('Usage:', sys.argv[0], 'server', 'prefix')
+        sys.exit(1)
+    server = sys.argv[1]
+    prefix = sys.argv[2]
+    print('Finding tests prefixed with {} at server {}'.format(prefix, server))
```

Comment:
FYI, let's stick with % or .format consistently within the file (so please translate my previous % suggestions to .format)